NC | NSFS | Versioning | Put-Object request for folder is failing after second time

### Environment info
- NooBaa Version: **VERSION**
- Platform: **Kubernetes 1.14.1 | minikube 1.1.1 | OpenShift 4.1 | other: specify**

### Actual behavior
1. When trying to push the object into a folder key(a/b/c/), the first two times it works but the third time onwards request fails with an internal error. 

### Expected behavior
1. customer should be able to push data to the folder key multiple times without issues.
```
aws s3api put-object --bucket bucket-v --key /a/b/c/ --body test-version.txt 
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings

An error occurred (InternalError) when calling the PutObject operation (reached max retries: 2): We encountered an internal error. Please try again.
```

### Steps to reproduce
1. create a bucket and enable versioning for that bucket 
```
aws s3 mb s3://bucket-v
aws s3api put-bucket-versioning --bucket bucket-v --versioning-configuration Status=Enable
```
2. call put-object three times.
```
aws s3api put-object --bucket bucket-v --key /a/b/c/ --body test-data.txt 
````
Third time request will fail

### More information - Screenshots / Logs / Other output

Log :-
```
Oct-28 13:09:00.264 [Upgrade/43254]    [L0] core.endpoint.s3.ops.s3_put_object:: PUT OBJECT bucket-v /a/b/c/  
Oct-28 13:09:00.291 [Upgrade/43254]  [WARN] core.sdk.namespace_fs:: NamespaceFS._move_to_dest_version error: retries=9 should_retry=true new_ver_tmp_path=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/.noobaa-nsfs_671794d6d016fb26836f2975/uploads/8c3fc61e-694b-4b03-a85a-d93f43b17407 latest_ver_path=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/c/.folder [Error: File exists] { code: 'EEXIST', context: 'SafeLink _link_from.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/c/.folder _link_to.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/.versions/c_null _link_expected_mtime=1730100293304699904 _link_expected_inode=19937570 ' }
Oct-28 13:09:00.369 [Upgrade/43254]  [WARN] core.sdk.namespace_fs:: NamespaceFS._move_to_dest_version error: retries=8 should_retry=true new_ver_tmp_path=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/.noobaa-nsfs_671794d6d016fb26836f2975/uploads/8c3fc61e-694b-4b03-a85a-d93f43b17407 latest_ver_path=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/c/.folder [Error: File exists] { code: 'EEXIST', context: 'SafeLink _link_from.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/c/.folder _link_to.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/.versions/c_null _link_expected_mtime=1730100293304699904 _link_expected_inode=19937570 ' }
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

Oct-28 13:09:01.142 [Upgrade/43254]  [WARN] core.sdk.namespace_fs:: NamespaceFS._move_to_dest_version error: retries=0 should_retry=true new_ver_tmp_path=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/.noobaa-nsfs_671794d6d016fb26836f2975/uploads/8c3fc61e-694b-4b03-a85a-d93f43b17407 latest_ver_path=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/c/.folder [Error: File exists] { code: 'EEXIST', context: 'SafeLink _link_from.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/c/.folder _link_to.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/.versions/c_null _link_expected_mtime=1730100293304699904 _link_expected_inode=19937570 ' }
Oct-28 13:09:01.142 [Upgrade/43254] [EVENT]{"timestamp":"2024-10-28T07:39:01.142Z","host":"li-afa941cc-3474-11b2-a85c-fc0f61e15c5c.ibm.com","event":{"code":"noobaa_object_upload_failed","message":"Error while uploading object, value: /a/b/c/","description":"Upload object failed, error: Error: File exists","entity_type":"NODE","event_type":"ERROR","scope":"NODE","severity":"ERROR","state":"DEGRADED","arguments":{"bucket_path":"/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v","object_name":"/a/b/c/"},"pid":43254}}
Oct-28 13:09:01.143 [Upgrade/43254]  [WARN] core.sdk.namespace_fs:: NamespaceFS: upload_object buffer pool cleanup error [Error: File exists] { code: 'EEXIST', context: 'SafeLink _link_from.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/c/.folder _link_to.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/.versions/c_null _link_expected_mtime=1730100293304699904 _link_expected_inode=19937570 ' }
Oct-28 13:09:01.144 [Upgrade/43254] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/bucket-v//a/b/c/</Resource><RequestId>m2spjrj8-eiucm6-szw</RequestId></Error> PUT /bucket-v//a/b/c/ {"host":"localhost:6443","accept-encoding":"identity","user-agent":"aws-cli/2.15.48 Python/3.11.8 Linux/6.10.12-200.fc40.x86_64 exe/x86_64.fedora.40 prompt/off command/s3api.put-object","content-md5":"ntZm8vMQAqkXZICFVsecuQ==","expect":"100-continue","x-amz-date":"20241028T073900Z","x-amz-content-sha256":"UNSIGNED-PAYLOAD","authorization":"AWS4-HMAC-SHA256 Credential=y4GJmmgB0PPNPvNeB2Hz/20241028/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=5af1dc624a46079968d145fd42e50345349c6069495b0c695027defb64bc188f","content-length":"57"} Error: File exists - context: SafeLink _link_from.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/c/.folder _link_to.c_str()=/home/naveenpaul/WS/standalon/nsfs_root/nested_key/bucket-v/a/b/.versions/c_null _link_expected_mtime=1730100293304699904 _link_expected_inode=19937570

```

--- a/src/sdk/namespace_fs.js
+++ b/src/sdk/namespace_fs.js
@@ -1388,7 +1388,7 @@ class NamespaceFS {
         dbg.log1('NamespaceFS._finish_upload:', open_mode, file_path, upload_path, fs_xattr);
 
         if (!same_inode && !part_upload) {
-            await this._move_to_dest(fs_context, upload_path, file_path, target_file, open_mode, params.key);
+            await this._move_to_dest(fs_context, upload_path, file_path, target_file, open_mode, params.key, is_dir_content);
         }
 
         // when object is a dir, xattr are set on the folder itself and the content is in .folder file
@@ -1421,13 +1421,16 @@ class NamespaceFS {
     }
 
     // move to dest GPFS (wt) / POSIX (w / undefined) - non part upload
-    async _move_to_dest(fs_context, source_path, dest_path, target_file, open_mode, key) {
+    async _move_to_dest(fs_context, source_path, dest_path, target_file, open_mode, key, is_dir_content) {
+        dbg.log2('_move_to_dest', fs_context, source_path, dest_path, target_file, open_mode, key, is_dir_content);
         let retries = config.NSFS_RENAME_RETRIES;
         // will retry renaming a file in case of parallel deleting of the destination path
         for (;;) {
             try {
                 await native_fs_utils._make_path_dirs(dest_path, fs_context);
-                if (this._is_versioning_disabled()) {
+                if (this._is_versioning_disabled() ||
+                    (this._is_versioning_enabled() && is_dir_content)) {
+                // dir_content is not supported in versioning, hence we will treat it like versioning disabled
                     if (open_mode === 'wt') {
                         await target_file.linkfileat(fs_context, dest_path);
                     } else {

--- a/src/test/unit_tests/test_bucketspace_versioning.js
+++ b/src/test/unit_tests/test_bucketspace_versioning.js
@@ -339,6 +339,21 @@ mocha.describe('bucketspace namespace_fs - versioning', function() {
                 }
                 assert.deepEqual(res_version_ids, res_put_version_ids);
             });
+
+            // dir_content is not supported in versioning, but we want to make sure there are no errors
+            mocha.it('put object - versioning enabled - directory content', async function() {
+                await s3_uid6.putBucketVersioning({ Bucket: nested_keys_bucket_name, VersioningConfiguration: { MFADelete: 'Disabled', Status: 'Enabled' } });
+                const key_as_dir_content = '/a/b/c/';
+                const size = 4; // in the original issue the error started on the 3rd PUT of dir content
+                const res_put_object = [];
+                for (let i = 0; i < size; i++) {
+                    const res = await s3_uid6.putObject({ Bucket: nested_keys_bucket_name, Key: key_as_dir_content, Body: body1 });
+                    res_put_object.push(res);
+                }
+                assert(res_put_object.length === size);
+                const check_version_id_exists = res_put_object.every(res => res.VersionId !== undefined);
+                assert.ok(check_version_id_exists);
+            });
         });
 
         mocha.describe('mpu object', function() {

